### DISCUSSION TITLE
在stm32f1 使用 tinyusb 的时候 报错

### DISCUSSION THREAD

**Discussion by changpan1110** [hathach/tinyusb#2545] [General]
packages\TinyUSB-v0.14.0\src\common/tusb_compiler.h(134): error:  #59: function call is not allowed in a constant expression 提示这样
![111111111111111111111111111111111](https://github.com/hathach/tinyusb/assets/46922417/b863d406-7e8a-408b-a4cb-01a3156ec4b1)
![222222222222222222222](https://github.com/hathach/tinyusb/assets/46922417/bdefc1cc-c3b8-43b2-abec-841ee8d17375)
这样的话，怎么解决呢？

**Comment by HiFiPhile** [comment]
你可以先把TinyUSB替换到最新版,0.14太老了.

**Reply by changpan1110** [reply]
RTthread 最新的包是0.14.1版本 换成这个也是一样报错呢！

**Comment by HiFiPhile** [comment]
应该是Keil不识别`__has_attribute`, 但该行代码已存在很长时间，去年有人测试Keil也未遇到问题，是不是Keil版本过老？

**Comment by changpan1110** [comment]
![999999999999999999999](https://github.com/hathach/tinyusb/assets/46922417/3d7748b6-0991-4159-af55-aedcc910c135)
这是我的版本呢！
![image](https://github.com/hathach/tinyusb/assets/46922417/9b3c9c4a-0dc6-4798-860b-c8b2d306fb4a)
这是最新的版本

**Comment by HiFiPhile** [comment]
搞忘了，你把这块代码改一下试试：
https://github.com/hathach/tinyusb/blob/5738757e2ca73ab003df5bfbf5ec7b60850ef88b/src/common/tusb_compiler.h#L143C1-L151C9
```
#if __GNUC__ < 5
  #define TU_ATTR_FALLTHROUGH         do {} while (0)  /* fallthrough */
#elif defined(__has_attribute)
  #if __has_attribute(__fallthrough__)
    #define TU_ATTR_FALLTHROUGH         __attribute__((fallthrough))
  #else
    #define TU_ATTR_FALLTHROUGH         do {} while (0)  /* fallthrough */
  #endif
#else
  #define TU_ATTR_FALLTHROUGH         do {} while (0)  /* fallthrough */
#endif
```

**Comment by changpan1110** [comment]
把代码改成下面那样得话，感觉和
#define TU_ATTR_FALLTHROUGH         do {} while (0)  /* fallthrough */只留这句话效果一样呢
现在得tinyusb 最新得版本是0.14 还是0.14.1呢？




Highone
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "hathach/tinyusb"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年4月4日(星期四) 晚上11:23
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [hathach/tinyusb] 在stm32f1 使用 tinyusb 的时候 报错 (Discussion #2545)





 
搞忘了，你把这块代码改一下试试：
 https://github.com/hathach/tinyusb/blob/5738757e2ca73ab003df5bfbf5ec7b60850ef88b/src/common/tusb_compiler.h#L143C1-L151C9
 #if __GNUC__ < 5   #define TU_ATTR_FALLTHROUGH         do {} while (0)  /* fallthrough */ #elif defined(__has_attribute)   #if __has_attribute(__fallthrough__)     #define TU_ATTR_FALLTHROUGH         __attribute__((fallthrough))   #else     #define TU_ATTR_FALLTHROUGH         do {} while (0)  /* fallthrough */   #endif #else   #define TU_ATTR_FALLTHROUGH         do {} while (0)  /* fallthrough */ #endif

**Reply by HiFiPhile** [reply]
@Rbb666 你们试试升级到最新的master

**Comment by changpan1110** [comment]
最新的只有0.14.1&nbsp;
然后升级到最新的后 烧录到板子上 会出来 一个为止的usb设备



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2024年4月25日(周四) 晚上7:53
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [hathach/tinyusb] 在stm32f1 使用 tinyusb 的时候 报错 (Discussion #2545)




 
@Rbb666 你们试试升级到最新的master
